Extensions/RosettaTorch/Includes/AlphaZero/SelfPlay/SelfPlayer.hpp

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) 2019 Chris Ohk, Youngjoong Kim, SeungHyun Jeon
+
+// We are making my contributions/submissions to this project solely in our
+// personal capacity and are not conveying any rights to any intellectual
+// property of any third parties.
+
+// It is based on peter1591's hearthstone-ai repository.
+// References: https://github.com/peter1591/hearthstone-ai
+
+#ifndef ROSETTASTONE_TORCH_ALPHA_ZERO_SELF_PLAY_SELF_PLAYER_HPP
+#define ROSETTASTONE_TORCH_ALPHA_ZERO_SELF_PLAY_SELF_PLAYER_HPP
+
+#include <Agents/MCTSAgent.hpp>
+#include <AlphaZero/Logger/ILogger.hpp>
+#include <AlphaZero/SelfPlay/RunOptions.hpp>
+#include <AlphaZero/SelfPlay/RunResult.hpp>
+#include <AlphaZero/Training/TrainingData.hpp>
+#include <Judges/JSON/Recorder.hpp>
+#include <Judges/Judger.hpp>
+#include <NeuralNet/NeuralNetwork.hpp>
+
+#include <Rosetta/Commons/DeckCode.hpp>
+
+#include <chrono>
+
+namespace RosettaTorch::AlphaZero::SelfPlay
+{
+//!
+//! \brief AgentCallback class.
+//!
+//! This class is an agent callback for self-play of AlphaZero.
+//!
+class AgentCallback
+{
+ public:
+    //! Constructs agent callback with given \p logger.
+    //! \param logger The logger to write the log.
+    AgentCallback(ILogger& logger);
+
+    //! Processes something before calling Think() method.
+    //! \param view The board ref view.
+    void BeforeThink(const RosettaStone::BoardRefView& view);
+
+    //! Processes something related to agent.
+    //! \param view The board ref view.
+    //! \param iteration The number of iteration.
+    void Think(const RosettaStone::BoardRefView& view, std::uint64_t iteration);
+
+    //! Processes something after calling Think() method.
+    //! \param iteration The number of iteration.
+    void AfterThink(std::uint64_t iteration);
+
+ private:
+    ILogger& m_logger;
+    std::chrono::steady_clock::time_point m_nextShow;
+    std::chrono::milliseconds m_showInterval;
+};
+
+//!
+//! \brief SelfPlayer class.
+//!
+//! This class is a helper for self-playing the game that provides BeforeRun(),
+//! Run() and AfterRun().
+//!
+class SelfPlayer
+{
+ public:
+    //! Constructs self player with given \p logger.
+    //! \param logger The logger to write the log.
+    SelfPlayer(ILogger& logger);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Class 'SelfPlayer' has a constructor with 1 argument that is not explicit.](https://app.codacy.com/app/utilForever/RosettaStone/pullRequest?prid=4565217)